### PR TITLE
DEP: stats.mode: raise with non-numeric input

### PR DESCRIPTION
**Pull request by mdhaber**
#### Reference issue
https://github.com/scipy/scipy/issues/18254#issuecomment-1502436548

#### What does this implement/fix?
One resolution of gh-18245 was that `stats.mode` should raise when input is non-numeric. This implements that.

### CODE REVIEW

File: scipy/stats/_stats_py.py [hunk 1]
```diff
@@ -501,6 +501,12 @@ def mode(a, axis=0, nan_policy='propagate', keepdims=False):
 
     """  # noqa: E501
     # `axis`, `nan_policy`, and `keepdims` are handled by `_axis_nan_policy`
+    if not np.issubdtype(a.dtype, np.number):
+        message = ("Argument `a` is not recognized as numeric. "
```
Comment (mdhaber) [inline scipy/stats/_stats_py.py:505]: Phrased like this to reduce confusion when the input appears to be numeric but isn't coerced to numeric type (e.g. `Series` with pandas integers).

**Comment by j-bowhay** [comment]
I think there are some more places in the test to be lined up `test_stats.py`:
https://github.com/scipy/scipy/blob/c62be9c695f3195362862bc9e9328057bf390b86/scipy/stats/tests/test_stats.py#L2295-L2299
https://github.com/scipy/scipy/blob/c62be9c695f3195362862bc9e9328057bf390b86/scipy/stats/tests/test_stats.py#L2212

File: scipy/stats/_stats_py.py (hunk 1 above)
Comment (rgommers) [reply scipy/stats/_stats_py.py:505]: are you intending to drop `bool` on purpose here?

Comment (mdhaber) [reply scipy/stats/_stats_py.py:505]: Yes, for now. The deprecation warning was produced under the same conditions as the error message would be with this PR, so it should be OK.

**Review by j-bowhay — APPROVED** [review]

**Pull request merged**